### ISSUE TITLE
GemStone on Container-Based Infrastructure

### ISSUE THREAD

**Issue by fniephaus** [hpi-swa/smalltalkCI#28]
GemStone builds currently require builderCI fallback and don't run on Travis' container-based infrastructure, because setting up a GemStone VM requires sudo.
We should be able to set up a GemStone VM on the faster infrastructure by moving the setup code to:
https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/script/smalltalk.rb
The steps necessary to set up a GemStone VM can be found [here](https://github.com/dalehenrich/builderCI/blob/3950dc4c7e01dd5163a7265f8ce5d1e36d3cae69/installGemstone.sh#L17-L27). Obviously, a GemStone VM should only be set up for GemStone builds.

Maybe @dalehenrich can help with this. Are all steps in [here](https://github.com/dalehenrich/builderCI/blob/3950dc4c7e01dd5163a7265f8ce5d1e36d3cae69/installGemstone.sh#L17-L27) really necessary or is there an updated version of the installer?

**Comment by dalehenrich** [comment]
@fniephaus, the latest information about OS prereqs can be found [here](https://github.com/GsDevKit/GsDevKit_home/blob/dev/bin/utils/installOsPrereqs).

In fact *starting from scratch*, I think that the gemstone support should be based on [GsDevKit_home](https://github.com/GsDevKit/GsDevKit_home/tree/dev) (note I'm referencing the dev branch) ... the scripts in GsDevKit_home are written to install GemStone (including download from ftp server) and there are scripts for creating stones etc, that are much more flexible than the old **builderCI** scripts -- including OS X support. I was considering updating the **builderCI** scripts, but starting fresh is probably a much better solution ... 

Along these lines, would it be possible to pre-install the GemStone releases in the smalltalk container?

**Comment by fniephaus** [comment]
Thanks for the info, @dalehenrich.
Since we are able to use sudo in the [`smalltalk.rb`](https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/script/smalltalk.rb), we just have to do all the magic there if `smalltalk` is set to `GemStone*`.

**Comment by dalehenrich** [comment]
I've started work on the Gemstone-specific bits based on [GsDevKit_home](https://github.com/GsDevKit/GsDevKit_home/tree/dev) over [here](https://github.com/dalehenrich/smalltalkCI/tree/gs_master) ... so if you guys could add the system configuration bits to smalltalk.rb I'll be able to start testing GemStone in the containers ...

**Comment by fniephaus** [comment]
Wow, that escalated quickly :smile: 

I'd also like to get this done asap, but getting the `smalltalk.rb` updated can take a while, especially when we need to run more advanced commands that require sudo.
Unfortunately, I'm afraid we won't be able to just call the [installOsPrereqs](https://github.com/GsDevKit/GsDevKit_home/blob/dev/bin/utils/installOsPrereqs) script within the smalltalk.rb, because that could potentially be a security problem for Travis. Instead, we should probably add the missing packages for Ubuntu (only for GemStone builds) and then set up the shared memory.
We literally have no way to test the code before it hits production, so we really need to make sure not to miss or break anything.

Nonetheless, let me know if you want your code to be merged, so you can start testing it using `sudo: true`.

**Comment by dalehenrich** [comment]
Haha ... nothing like striking when the iron is hot ... it looks like `sudo: true` does the trick and I really like the new structure! Good effort!

**Comment by dalehenrich** [comment]
As an additional requirement that is unique to travis-ci, [i found that we need to do the following](https://github.com/GsDevKit/GsDevKit_home/issues/50) to get gemstone to run correctly:

```yml
addons:
  hosts:
    - travis.dev

install:
  - sudo hostname travis.dev
```

The `sudo hostname` is needed, because `hostname` does not use `/etc/hosts` and GemStone uses the same C api that `hostname` uses **and** requires that the name returned by `hostname` is in `/etc/hosts`...

**Comment by dalehenrich** [comment]
Minor milestone - [tests green running GemStone via GsDevKit_home in a container --- with sudo: true](https://travis-ci.org/dalehenrich/smalltalkCI)

**Comment by fniephaus** [comment]
Nice! If you open a PR I can merge your code in if you want.

BTW: with `sudo: true` you explicitly disable the new container-based infrastructure, but once we have updated the `smalltalk.rb` it should be possible to just do `sudo: false` to enable it.

**Comment by dalehenrich** [comment]
I don't think I'm quite ready to merge as I've had to monkey with 
.travis.yml files on master and gs_master and of course I have to do the 
`sudo hostname travis.dev` and I am just now seeing if I can push these 
settings to the fixture .travis.yml (I'm using a gs_fixture branch with 
my changes)... but then that would mean the the gemstone projects 
couldn't use the same fixture as the other projects (unless the `sudo 
hostname travis.dev` install can be conditional on GemStone ...

I'm also considering that I may want to include GsDevKit-* and 
GemStone-* as targets ... I need to be able to test GsDevKit with 
smalltalkCI and bypass the GsDevKit-* install ---- or spend more time 
developing scripts :)

The good news is that I've got GemSTone itself running in the container 
... now I just have to figure out how to merge the changes I've made 
back into the master :)

**Comment by fniephaus** [comment]
No worries. Please take your time while we figure how to prepare a container-based image for a GemStone vm...and of course let me know if there's anything I can help with.

**Comment by fniephaus** [comment]
@bahnfahren, @chistopher, @jchromik, and @Nef10:
I am assuming you are busy working on [squeak-parable](https://github.com/HPI-BP2015H/squeak-parable/) and maybe also #23, so I will have a look at how to prepare an image for a GemStone vm.

**Comment by dalehenrich** [comment]
osx has a different method for setting the hostname .. here's some bash for both linux and osx (in addition to the `addons:host:travis.dev`):

```
  if [ "${TRAVIS_OS_NAME}" = "linux" ] ; then
    sudo hostname travis.dev
  else # then osx
    sudo scutil --set HostName travis.dev
  fi
```

**Comment by fniephaus** [comment]
Yay: https://travis-ci.org/hpi-swa/smalltalkCI/builds/100495992

**Comment by dalehenrich** [comment]
Excellent!

**Comment by dalehenrich** [comment]
I will start converting a few projects to smalltalkCI as a I touch them and then update the wiki and read with more current info ...

**Comment by fniephaus** [comment]
Now that we have a first GemStone build running, I am wondering what other versions smalltalkCI can and should support. What do you think, @dalehenrich?

**Comment by dalehenrich** [comment]
You mean other versions of GemStone? The range runs from 2.4.4.1 to 
3.2.12 and all of them ought to work right out of the box ...

**Comment by fniephaus** [comment]
Yes, GemStone versions. Are there some major ones that we should add to smalltalkCI's build matrix?

**Comment by dalehenrich** [comment]
2.4.7, 3.1.0.6, and 3.2.12 should be good enough for now ....

**Comment by fniephaus** [comment]
Mhh I am getting this error now:

>Error: while installing tODE on server: 'a UserDefinedError occurred (error 2318), reason:halt, Error during download: ''UserDefinedError: API rate limit exceeded for 208.78.110.199. (But here''''s the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)''. Error could have been caused by: ''

e.g.: https://travis-ci.org/hpi-swa/smalltalkCI/jobs/100713275

Is there a way we can authenticate these requests?

**Comment by fniephaus** [comment]
This appears to be the [GitHub API rate limit](https://developer.github.com/v3/#rate-limiting). Not sure what we could do about this...when loading code from GitHub, Metacello will not just download but clone the entire repo, right? The problem would disappear if it would only download the repo, it's also unlikely that the user is interested in the repo history, right?

**Comment by dalehenrich** [comment]
Yes this is the github API rate limit ... when Metacello is trying to resolve a tag pattern, it makes requests against the github api, and is subject to the api rate limit ... authentication is possible, but it does require someone's credentials and I'm not sure you can specify credentials in a public repository ... If you figure out how to do that let me know ...  I think there are some facilities for setting secure env vars and there's a smalltalk api for setting username and password in an image, but it only pops up occasionally and waiting a bit and trying again seems to fill the bill ... I get hit most often when slamming travis while trying to debug a particular issue :) ... also seems to happen when sharing an ip address with another unrelated build is hitting the github api ...

**Comment by fniephaus** [comment]
There are a couple of ways to [encrypt keys on Travis](https://docs.travis-ci.com/user/encryption-keys/), but they don't work across GitHub projects, because the encrypted data is tied to the corresponding project.
However, GitHub supports [read-only keys](https://developer.github.com/guides/managing-deploy-keys/#deploy-keys) for repositories that could be used. But I am not sure how feasible this approach is. 
Would we need to have a read-only key for each GitHub repo that holds Smalltalk code?
Where would we store these keys?

**Comment by dalehenrich** [comment]
I think that individual repo owners can decide whether or not they want to define keys ... and I haven't had the bandwidth to take this any further, since the failures I get due to read limits are not that frequent and restarting one or two jobs because of random network problems is pretty much common anyway:)

**Comment by dalehenrich** [comment]
I would think that the energy should be spent on created a cache of the 
build, then the only exposure to the rate limit is from the project 
itself (almost)

**Comment by fniephaus** [comment]
That's a good point. And for that we only need to cache the build folder which already includes `package-cache` and `github-cache`. The `.image` and `.changes` can be removed before that to reduce the cache size.

**Comment by fniephaus** [comment]
@dalehenrich do you mind looking through this: https://github.com/fniephaus/travis-build/commit/31be289bbf89178e9839f4b4e894072330bcdc92
This is a first version and I still need to add tests.

**Comment by fniephaus** [comment]
@dalehenrich this link is probably better: [`smalltalk.rb`](https://github.com/fniephaus/travis-build/blob/fniephaus-smalltalkCI/lib/travis/build/script/smalltalk.rb).

**Comment by fniephaus** [comment]
@dalehenrich Mh, `kern.sysv.mseg` does not seem to exist (anymore?).
https://github.com/GsDevKit/GsDevKit_home/blob/555475f48bca2960644a128f59e4567d42cc49f5/bin/downloadGemStone#L239
https://travis-ci.org/GsDevKit/GsDevKit_home/jobs/100163003#L303

Travis has deployed the patch to their staging instance and I now need to fix the [hostname issue](https://staging.travis-ci.org/fniephaus/SWT-Demo/jobs/475845#L1018). I will just add the current hostname to the `/etc/hosts` as you mentioned in https://github.com/hpi-swa/smalltalkCI/issues/38#issuecomment-172066733 without changing the current hostname. Is that correct?
Also, there currently is another problem that keeps Linux builds from installing the dependencies. I'm working on it...

**Comment by dalehenrich** [comment]
Thx @fniephaus ... created issue and will address in next release of GsDevKit_home ... the missing param doesn't appear to affect operation of GemStone on OSX ....

Hopefully adding current `hostname` to /etc/hosts should make GemStone happy .... as long as travis remains happy:)

**Comment by dalehenrich** [comment]
Oh ... what is the installation issue?

**Comment by fniephaus** [comment]
It's just another mistake that I've made. `travis-build` is a Ruby program that generates a bash script that then runs on a Travis worker. And I mixed up Ruby with Bash code...
I'm afraid we'll need to wait for the next update :cry:

**Comment by dalehenrich** [comment]
Ah okay ... no problem! It is well worth the wait :)

**Comment by fniephaus** [comment]
I hope so!

In the meantime, would you mind having a look at the code in [`dev`](https://github.com/hpi-swa/smalltalkCI/tree/dev) and check if this would work in GemStone?

The companion is located at: [`lib/SmalltalkCI-Core.st`](https://github.com/hpi-swa/smalltalkCI/blob/dev/lib/SmalltalkCI-Core.st).
The tests for the companion are located at https://github.com/hpi-swa/smalltalkCI/tree/dev/tests

Basically, `SmalltalkCI-Core` needs to be loaded into an image and then you can execute a CI run with:
`SmalltalkCISpec runCIFor: '/path/to/project/smalltalk.ston'`

**Comment by dalehenrich** [comment]
I'm kinda tied up at the moment ... I assume you expect the SmalltalkCI-Core to be present in an image, or do you load it as part of the startup?

If it needs to be pre-loaded a filetree package would be more convenient for GemStone and I'd just make the package part of the bootstrap step --- installed by https://github.com/GsDevKit/gsUpgrader ...

**Comment by fniephaus** [comment]
No worries, take your time. We need to wait for Travis anyway...

I am currently trying to do this (obvious not working at all at the moment):
https://github.com/hpi-swa/smalltalkCI/blob/dev/gemstone/run.sh#L80-L83

**Comment by dalehenrich** [comment]
Ah, GemStone interprets env vars at the beginning of file paths, so the following should work: '$SMALLTALK_CI_HOME/lib/SmalltalkCI-Core.st'; if `SMALLTALK_CI_HOME` is exported to the shell, before the installation of GemStone is started ... if the other path is used as a file path, the same trick should work ...

**Comment by fniephaus** [comment]
The entire EOF expression will already become something like this:
```smalltalk
(BinaryOrTextFile openReadOnServer: '/home/travis/...../lib/SmalltalkCI-Core.st') 
        fileIn;
        close.
SmalltalkCISpec runCIFor: '/home/travis/project/...../smalltalk.ston'
System commitTransaction.
```
So that should be fine, right?

**Comment by dalehenrich** [comment]
Yeah that looks pretty good, probably need to look up the `SmalltalkCISpec` class:

```Smalltalk
(BinaryOrTextFile openReadOnServer: '/home/travis/...../lib/SmalltalkCI-Core.st') 
        fileIn;
        close.
(Smalltalk at: #SmalltalkCISpec) runCIFor: '/home/travis/project/...../smalltalk.ston'
System commitTransaction.
```

**Comment by fniephaus** [comment]
:+1: noted!

**Comment by fniephaus** [comment]
I have opened https://github.com/travis-ci/travis-build/pull/607 which hopefully fixes all remaining issues that keep us from running GemStone tests on Travis containers.

**Comment by dalehenrich** [comment]
Looking forward to this --- great!

**Comment by dalehenrich** [comment]
I just found a spot where using [$GS_HOME/bin/.gsdevkitSysSetup](https://github.com/hpi-swa/smalltalkCI/blob/master/gemstone/run.sh#L38) may be bypassing some important installation ... so there's a chance that more work will be required for GemStone to function correctly once the prereqs and host name issues have been resolved

**Comment by fniephaus** [comment]
What else needs to be done? Maybe there's still time to squeeze this into the current PR.

**Comment by dalehenrich** [comment]
Nothing in the ruby script ... the fix can be made in the GemStone `run.sh` script ... but I won't fool with that until we get it running on the container (the problem is that we are creating the file `.gsdevkit...` to bypass `sudo` operations and it may bypass more than just `sudo` calls ... we just may never have gotten far enough to expose this particular issue :)

**Comment by fniephaus** [comment]
Ah fair enough, sounds good!

**Comment by fniephaus** [comment]
Closing this in favor of #51
